mam/v1/mam.go

```diff
@@ -0,0 +1,298 @@
+}
+
+// PayloadMinLength computes the minimum length of a payload
+//
+//	messageLength is the length of the message
+//	merkleTreeLength is the length of the merkle tree
+//	index is the leaf index
+//	security is the security level used to generate addresses
+//
+// Returns the minimum length of the payload
+func PayloadMinLength(messageLength, merkleTreeLength, index uint64, security SecurityLevel) uint64 {
+	siblingNumber := MerkleDepth(merkleTreeLength/HashTrinarySize) - 1
+	return EncodedLength(int64(index)) + EncodedLength(int64(messageLength)) +
+		HashTrinarySize + messageLength + HashTrytesSize +
+		uint64(security)*ISSKeyLength + EncodedLength(int64(siblingNumber)) +
+		siblingNumber*HashTrinarySize
+}
+
+// MamCreate creates a signed, encrypted payload from a message.
+//
+//	payloadLength is the length of the payload
+//	message is the message to encrypt
+//	sideKey is the encryption key
+//	merkleTree is the merkle tree
+//	merkleTreeLength is the length of the merkle tree
+//	leafCount is the number of leaves of the merkle tree
+//	index is the leaf index
+//	nextRoot is the merkle root of the next MAM payload
+//	start is the offset used to generate addresses
+//	seed is the seed used to generate addresses - Not sent over the network
+//	security is the security level used to generate addresses
+//
+// Returns:
+//  payload is the payload of the encrypted message
+//  payloadMinLength is the length of the payload needed to encrypt the message
+//	err is the error message
+func MamCreate(payloadLength uint64,
+	message Trytes, sideKey Trytes,
+	merkleTree Trits, merkleTreeLength uint64,
+	leafCount uint64, index uint64,
+	nextRoot Trits, start uint64,
+	seed Trytes, security SecurityLevel) (payload Trits, payloadMinLength uint64, err error) {
+
+	if security > 3 {
+		return nil, 0, fmt.Errorf("invalid security %v", security)
+	}
+
+	messageTrits, err := TrytesToTrits(message)
+	if err != nil {
+		return nil, 0, err
+	}
+	messageLength := len(messageTrits)
+
+	sideKey = Pad(sideKey, 81)
+	sideKeyTrits, err := TrytesToTrits(sideKey)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	siblingsNumber := MerkleDepth(merkleTreeLength/HashTrinarySize) - 1
+
+	indexTrits, encIndexLength, err := EncodeInt64(int64(index))
+	if err != nil {
+		return nil, 0, err
+	}
+
+	messageLengthTrits, encMessageLengthLenght, err := EncodeInt64(int64(messageLength))
+	if err != nil {
+		return nil, 0, err
+	}
+
+	siblingsNumberTrits, encSiblingsNumberLength, err := EncodeInt64(int64(siblingsNumber))
+	if err != nil {
+		return nil, 0, err
+	}
+
+	signatureLength := uint64(security) * ISSKeyLength
+	payloadMinLength = encIndexLength + encMessageLengthLenght + HashTrinarySize +
+		uint64(messageLength) + HashTrytesSize + signatureLength +
+		encSiblingsNumberLength + (siblingsNumber * HashTrinarySize)
+
+	if payloadLength < payloadMinLength {
+		return nil, 0, fmt.Errorf("payload too short: needed %d, given %d", payloadMinLength, payloadLength)
```

Comment:
Make an explicit `Err...` variable at the start of the file and use `errors.Wrapf(Err..., "needed %d, given %d", minLength, payloadLength). So people can actually type switch on the origin error.